sdk/provisioning/Azure.Provisioning/src/ProvisioningDeployment.cs

```diff
@@ -0,0 +1,134 @@
+    /// whether the value is considered secure.
+    /// </summary>
+    public IReadOnlyDictionary<string, object?> Outputs { get; }
+    // TODO: Do we want to replace this with IReadOnlyDict<string, BicepOutput>
+    // to make it harder to misuse?
+
+    /// <summary>
+    /// Creates a new <see cref="ProvisioningDeployment"/>.
+    /// </summary>
+    /// <param name="context">
+    /// The provisioning context that was used to deploy the resources.
+    /// </param>
+    /// <param name="deployment">
+    /// The <see cref="ArmDeploymentResource"/> that was used to deploy
+    /// the infrastructure template.
+    /// </param>
+    /// <param name="outputs">
+    /// A dictionary of template outputs from the deployment.
+    /// </param>
+    /// <exception cref="ArgumentException">
+    /// Thrown when the deployment resource doesn't have its Data populated.
+    /// </exception>
+    internal ProvisioningDeployment(ProvisioningContext context, ArmDeploymentResource deployment, IReadOnlyDictionary<string, object?> outputs)
+    {
+        Context = context;
+        Deployment = deployment.HasData ? deployment :
+            throw new ArgumentException($"The {nameof(deployment)} must have its {nameof(ArmDeploymentResource.Data)} property set.", nameof(deployment));
+        Outputs = outputs;
+    }
+
+    /// <summary>
+    /// Create a data-plane client for a specific Azure resource.
+    /// </summary>
+    /// <typeparam name="TClient">
+    /// The type of client to create.  This can be inferred automatically from
+    /// a resource implementing <see cref="IClientCreator{TClient, TOptions}"/>.
+    /// </typeparam>
+    /// <typeparam name="TOptions">
+    /// The type of <see cref="ClientOptions"/> used to customize creation.
+    /// This can be inferred automatically from a resource implementing
+    /// <see cref="IClientCreator{TClient, TOptions}"/>.
+    /// </typeparam>
+    /// <param name="resource">
+    /// A resource that was provisioned as part of this deployment.
+    /// </param>
+    /// <param name="credential"></param>
+    /// <param name="options"></param>
+    /// <returns></returns>
+    /// <exception cref="InvalidOperationException">
+    /// Throw when trying to create a client from a failed deployment.
+    /// </exception>
+    public TClient CreateClient<TClient, TOptions>(IClientCreator<TClient, TOptions> resource, TokenCredential? credential = default, TOptions? options = default)
+        where TOptions : ClientOptions
+    {
+        if (ProvisioningState != ArmProvisioningState.Succeeded)
+        {
+            throw new InvalidOperationException($"Cannot create a client because the deployment did not succeed: {Error}");
+        }
+        credential ??= Context.DefaultClientCredential;
+        if (options is not null) { Context.ConfigureClientOptionsCallback?.Invoke(options); }
+        return resource.CreateClient(this, credential, options);
+    }
```

Comment:
Or you can go end to end with `BlobServiceClient dataPlane = deployment.CreateClient(storageAcct);` and have it fish out the correct outputs and instantiate a client for you.